File: src/utils/logger.py
import logging
from collections.abc import Callable

from rich.console import Console
from rich.logging import RichHandler

FORMAT = "%(message)s"

logging.basicConfig(
    level=logging.INFO,
    format="%(message)s",
    datefmt="[%X]",
    handlers=[RichHandler(rich_tracebacks=True)],
)

DEFAULT_LOG_LEVEL_MAP = {
    "critical": True,
    "error": True,
    "warning": True,
    "info": True,
    "debug": True,
}


# Decorator to stringify the message
def stringify(func: Callable) -> Callable:
    def inner(self, method_type: str, *msg: object, sep=" ") -> None:
        nmsg = []
        for v in msg:
            if not isinstance(v, str):
                nmsg.append(str(v))
            else:
                nmsg.append(v)

        return func(self, method_type, *nmsg, sep=sep)

    return inner


class Logger:
    def __init__(
        self,
        name,
        message_format="%(message)s",
        date_format="[%X]",
    ) -> None:
        self.log = logging.getLogger(name)
        self.log.setLevel(logging.DEBUG)
        self.log.__format__ = message_format
        self.log.__date_format__ = date_format
        self.reset_log_levels()

    def set_log_levels(self, show_logs_by_type) -> None:
        self.show_logs_by_type = {**DEFAULT_LOG_LEVEL_MAP, **show_logs_by_type}

    def reset_log_levels(self) -> None:
        self.show_logs_by_type = DEFAULT_LOG_LEVEL_MAP

    def debug(self, *msg: object, sep=" ") -> None:
        return self.logutil("debug", *msg, sep=sep)

    def info(self, *msg: object, sep=" ") -> None:
        return self.logutil("info", *msg, sep=sep)

    def warning(self, *msg: object, sep=" ") -> None:
        return self.logutil("warning", *msg, sep=sep)

    def error(self, *msg: object, sep=" ") -> None:
        return self.logutil("error", *msg, sep=sep)

    def critical(self, *msg: object, sep=" ") -> None:
        return self.logutil("critical", *msg, sep=sep)

    # set stack level to 3 so that the caller of this function is logged, not this function itself.
    # stack-frame - self.log.debug - logutil - stringify - log method - caller
    @stringify
    def logutil(self, method_type: str, *msg: object, sep=" ") -> None:
        if self.show_logs_by_type[method_type] is False:
            return None
        logger_func = getattr(self.log, method_type, None)
        if not logger_func:
            msg = f"Logger has no method {method_type}"
            raise AttributeError(msg)
        return logger_func(sep.join(msg), stacklevel=4)


logger = Logger(__name__)
console = Console()

File: omrchecker-js/packages/core/src/utils/logger.ts
/**
 * Logger utility for OMRChecker.
 *
 * Migrated from src/utils/logger.py.
 * Wraps browser console methods with per-level filtering.
 */

const DEFAULT_LOG_LEVEL_MAP = {
  critical: true,
  error: true,
  warning: true,
  info: true,
  debug: true,
} as const;

type LogLevelMap = { [K in keyof typeof DEFAULT_LOG_LEVEL_MAP]?: boolean };

export class Logger {
  name: string;
  showLogsByType: Record<string, boolean>;

  constructor(name: string) {
    this.name = name;
    this.showLogsByType = { ...DEFAULT_LOG_LEVEL_MAP };
  }

  setLogLevels(levels: LogLevelMap): void {
    this.showLogsByType = { ...DEFAULT_LOG_LEVEL_MAP, ...levels };
  }

  resetLogLevels(): void {
    this.showLogsByType = { ...DEFAULT_LOG_LEVEL_MAP };
  }

  private logutil(methodType: string, ...msg: unknown[]): void {
    if (this.showLogsByType[methodType] === false) return;
    const str = msg.map(v => (typeof v === 'string' ? v : String(v))).join(' ');
    if (methodType === 'critical' || methodType === 'error') {
      console.error(str);
    } else if (methodType === 'warning') {
      console.warn(str);
    } else if (methodType === 'debug') {
      console.debug(str);
    } else {
      console.log(str);
    }
  }

  debug(...msg: unknown[]): void {
    this.logutil('debug', ...msg);
  }

  info(...msg: unknown[]): void {
    this.logutil('info', ...msg);
  }

  warning(...msg: unknown[]): void {
    this.logutil('warning', ...msg);
  }

  error(...msg: unknown[]): void {
    this.logutil('error', ...msg);
  }

  critical(...msg: unknown[]): void {
    this.logutil('critical', ...msg);
  }
}

export const logger = new Logger('omrchecker');
